File: src/main/java/dev/psygamer/wireframe/world/World.java
package dev.psygamer.wireframe.world;

import dev.psygamer.wireframe.block.Block;
import dev.psygamer.wireframe.block.state.BlockState;
import dev.psygamer.wireframe.internal.world.InternalWorld;
import dev.psygamer.wireframe.util.BlockPosition;

public interface World extends BlockReader {
	
	enum UpdateFlag {
		
		/**
		 * Causes neighbor updates to be sent to all surrounding blocks.
		 */
		NOTIFY_NEIGHBORS(1 << 0),
		
		/** Updates path-finding */
		BLOCK_UPDATE(1 << 1),
		
		/** Stops the blocks from being marked for a render update */
		NO_RERENDER(1 << 2),
		
		/**
		 * Makes the block be re-rendered immediately, on the main thread.
		 * If NO_RERENDER is set, then this will be ignored
		 */
		RERENDER_MAIN_THREAD(1 << 3),
		
		/**
		 * Causes neighbor updates to be sent to all surrounding blocks (including
		 * diagonals).
		 */
		UPDATE_NEIGHBORS(1 << 4),
		
		/**
		 * Prevents neighbor changes from spawning item drops, used by
		 */
		NO_NEIGHBOR_DROPS(1 << 5),
		
		/**
		 * Tell the block being changed that it was moved, rather than removed/replaced,
		 */
		IS_MOVING(1 << 6),
		
		/**
		 * Causes neighbor updates to be sent to all surrounding blocks
		 * and updates path-finding.
		 */
		DEFAULT(NOTIFY_NEIGHBORS.internalFlag | BLOCK_UPDATE.internalFlag),
		
		/**
		 * Causes neighbor updates to be sent to all surrounding blocks, updates path-finding
		 * and Makes the block be re-rendered immediately, on the main thread.
		 */
		DEFAULT_AND_RERENDER(DEFAULT.internalFlag | RERENDER_MAIN_THREAD.internalFlag);
		
		private final int internalFlag;
		
		UpdateFlag(final int internalFlag) {
			this.internalFlag = internalFlag;
		}
		
		public int getInternalFlag() {
			return this.internalFlag;
		}
	}
	
	static World get(final net.minecraft.world.World internalWorld) {
		if (internalWorld == null)
			return null;
		
		return new InternalWorld(internalWorld);
	}
	
	boolean isClientSide();
	
	boolean isServerSide();
	
	boolean setBlock(final Block block, final BlockPosition position);
	
	boolean setBlock(final Block block, final BlockPosition position, final UpdateFlag updateFlag);
	
	boolean setBlockState(final BlockState blockState, final BlockPosition position);
	
	boolean setBlockState(final BlockState blockState, final BlockPosition position, final UpdateFlag updateFlag);
	
	boolean isReplaceable(final BlockPosition position);
	
	void notifyNeighbours(final BlockPosition position);
	
	void notifyNeighbours(final BlockPosition position, Block block);
	
	boolean breakBlock(final BlockPosition position);
	
	boolean breakBlock(final BlockPosition position, final boolean dropItems);
	
	boolean isAir(final BlockPosition position);
	
	boolean isBlock(final Block block, final BlockPosition position);
	
	boolean isLoaded(final BlockPosition position);
	
	float getBlockLightLevel(final BlockPosition blockPosition);
	
	float getSkyLightLevel(final BlockPosition blockPosition);
	
	long getTime();
	
	long getTicks();
}

File: src/main/kotlin/dev/psygamer/wireframe/api/world/World.kt
package dev.psygamer.wireframe.api.world

import dev.psygamer.wireframe.api.block.*
import dev.psygamer.wireframe.util.BlockPosition

interface World : BlockReader {
	enum class UpdateFlag(val flag: Int) {
		/** Causes neighbor updates to be sent to all surrounding blocks. */
		NOTIFY_NEIGHBORS(0x1),

		/** Updates path-finding. */
		BLOCK_UPDATE(0x2),

		/** Stops the blocks from being marked for a render update. */
		NO_RENDER(0x4),

		/**
		 * Makes the block be re-rendered immediately, on the main thread.
		 * If [NO_RENDER] is set, then this will be ignored.
		 */
		RENDER_MAIN_THREAD(0x8),

		/** Causes neighbor updates to be sent to all surrounding blocks (including diagonals). */
		UPDATE_NEIGHBORS(0x10),

		/** Prevents neighbor changes from spawning item drops. */
		NO_NEIGHBOR_DROPS(0x20),

		/** Tells the block which is being changed that it was moved, rather than removed/replaced. */
		IS_MOVING(0x40),

		/** Causes neighbor updates to be sent to all surrounding blocks (including diagonals) and updates path-finding. */
		DEFAULT(NOTIFY_NEIGHBORS.flag or BLOCK_UPDATE.flag),

		/** Causes neighbor updates to be sent to all surrounding blocks (including diagonals),
		 * updates path-finding and makes the block be re-rendered immediately, on the main thread. */
		DEFAULT_AND_RENDER(DEFAULT.flag or RENDER_MAIN_THREAD.flag);

	}

	val isClientSide: Boolean
	val isServerSide: Boolean

	val dayTime: Long
	val gameTime: Long

	fun setBlock(block: Block, position: BlockPosition): Boolean
	fun setBlock(block: Block, position: BlockPosition, updateFlag: UpdateFlag): Boolean

	fun setBlockState(blockState: BlockState, position: BlockPosition): Boolean
	fun setBlockState(blockState: BlockState, position: BlockPosition, updateFlag: UpdateFlag): Boolean

	fun isReplaceable(position: BlockPosition): Boolean

	fun notifyNeighbours(position: BlockPosition)
	fun notifyNeighbours(position: BlockPosition, block: Block)

	fun breakBlock(position: BlockPosition): Boolean
	fun breakBlock(position: BlockPosition, dropItems: Boolean): Boolean

	fun isAir(position: BlockPosition): Boolean
	fun isLoaded(position: BlockPosition): Boolean
	fun isBlock(position: BlockPosition, block: Block): Boolean

	fun getBlockLightLevel(position: BlockPosition): Int
	fun getSkyLightLevel(position: BlockPosition): Int
}